test/tools/TestExe/TestExe.cs

```diff
@@ -41,6 +54,51 @@ private static int Main(string[] args)
             return 0;
         }
 
+        private static void WriteBase64Bytes(string value)
+        {
+            using Stream stdout = Console.OpenStandardOutput();
+            stdout.Write(Convert.FromBase64String(value));
+        }
+
+        private static void WriteBytes(ReadOnlySpan<string> args)
+        {
+            using Stream stdout = Console.OpenStandardOutput();
+            foreach (string arg in args)
+            {
+                if (!byte.TryParse(arg, NumberStyles.AllowHexSpecifier, provider: null, out byte value))
+                {
+                    throw new ArgumentException(
+                        nameof(args),
+                        "All args after -writebytes must be single byte hex strings.");
+                }
+
+                stdout.WriteByte(value);
+            }
+        }
+
+        [SkipLocalsInit]
+        private static void ReadBytes()
+        {
+            using Stream stdin = Console.OpenStandardInput();
+            Span<byte> buffer = stackalloc byte[0x200];
+            Unsafe.InitBlock(ref MemoryMarshal.GetReference(buffer), 0, 0x200);
+            Span<char> hex = stackalloc char[] { '\0', '\0' };
```

Comment:
Maybe a little intense for a test utility but it's still very readable. Definitely open to reverting to a less performance focused implementation if that's preferred.